File: com/jseidemann/adventofcode/day2/Day2Pt2.java
package com.jseidemann.adventofcode.day2;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.List;

public class Day2Pt2 {

    public static void main(String[] args) throws IOException {
        List<String> inputStrings = Files.readAllLines(Paths.get("com/jseidemann/adventofcode/day2/adventofcode_puzzle2_input.txt"));

        //@formatter:off
        int total = inputStrings.stream()
                .map(Round::new)
                .mapToInt(Round::getScore)
                .sum();
        //@formatter:on

        System.out.println("I got " + total + " points");
    }

    static class Round {
        OppPlay oppPlay;
        Result result;

        public Round(String input) {
            this.oppPlay = new OppPlay(input.substring(0, 1));
            this.result = Result.valueOf(input.substring(2, 3));
        }

        public Integer getScore() {
            return result.value + oppPlay.calculateMyPlay(result).value;
        }
    }

    static class OppPlay {
        String oppPlay;

        public OppPlay(String oppPlay) {
            this.oppPlay = oppPlay;
        }

        MyPlay calculateMyPlay(Result result) {
            if (oppPlay.equals("A")) {
                switch (result) {
                    case X: return MyPlay.C;
                    case Y: return MyPlay.A;
                    case Z: return MyPlay.B;
                    default: throw new RuntimeException();
                }
            } else if (oppPlay.equals("B")) {
                switch (result) {
                    case X: return MyPlay.A;
                    case Y: return MyPlay.B;
                    case Z: return MyPlay.C;
                    default: throw new RuntimeException();
                }
            } else {
                switch (result) {
                    case X: return MyPlay.B;
                    case Y: return MyPlay.C;
                    case Z: return MyPlay.A;
                    default: throw new RuntimeException();
                }
            }
        }
    }

    enum Result {
        X(0), Y(3), Z(6);

        Integer value;

        Result(Integer val) {
            this.value = val;
        }
    }

    enum MyPlay {
        A(1), B(2), C(3);

        Integer value;

        MyPlay(Integer val) {
            this.value = val;
        }
    }
}

File: com/jseidemann/adventofcode/day2/Day2Pt2.kt
package com.jseidemann.adventofcode.day2

import java.io.IOException
import java.nio.file.Files
import java.nio.file.Paths

object Day2Pt2 {
    @Throws(IOException::class)
    @JvmStatic
    fun main(args: Array<String>) {
        val inputStrings = Files.readAllLines(Paths.get("com/jseidemann/adventofcode/day2/adventofcode_puzzle2_input.txt"))

        //@formatter:off
        val total = inputStrings.stream()
                .map(::Round)
                .mapToInt(Round::calculateScore)
                .sum()
        //@formatter:on
        println("I got $total points")
    }

    internal class Round(input: String) {
        var oppPlay: OppPlay = OppPlay(input.substring(0, 1))
        var result: Result = Result.valueOf(input.substring(2, 3))
        fun calculateScore(): Int = result.value + oppPlay.calculateMyPlay(result).value

    }

    internal class OppPlay(var oppPlay: String) {
        fun calculateMyPlay(result: Result?): MyPlay {
            return when (oppPlay) {
                "A" -> {
                    when (result) {
                        Result.X -> MyPlay.C
                        Result.Y -> MyPlay.A
                        Result.Z -> MyPlay.B
                        else -> throw RuntimeException()
                    }
                }
                "B" -> {
                    when (result) {
                        Result.X -> MyPlay.A
                        Result.Y -> MyPlay.B
                        Result.Z -> MyPlay.C
                        else -> throw RuntimeException()
                    }
                }
                else -> {
                    when (result) {
                        Result.X -> MyPlay.B
                        Result.Y -> MyPlay.C
                        Result.Z -> MyPlay.A
                        else -> throw RuntimeException()
                    }
                }
            }
        }
    }

    internal enum class Result(var value: Int) {
        X(0), Y(3), Z(6);
    }

    internal enum class MyPlay(var value: Int) {
        A(1), B(2), C(3);
    }
}